### PR TITLE
Add unsaved changes message

### PR DESCRIPTION
**Pull request by phalkunz**
Issue: https://github.com/silverstripe/silverstripe-framework/issues/6683

### CODE REVIEW

File: client/src/boot/BootRoutes.js [hunk 1]
```diff
@@ -92,6 +94,22 @@ class BootRoutes {
       </ApolloProvider>,
       document.getElementsByClassName('cms-content')[0]
     );
+
+    history.listenBeforeUnload(() => {
+      if (this.shouldConfirmBeforeUnload()) {
+        return i18n._t('Admin.CONFIRMUNSAVEDSHORT');
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: Can you please use the default string as well?

`return i18n._t('Admin.CONFIRMUNSAVEDSHORT', 'WARNING: Your changes have not been saved.');`

File: client/src/boot/BootRoutes.js [hunk 2]
```diff
@@ -92,6 +94,22 @@ class BootRoutes {
...
+      }
+
+      return null;
+    });
+
+    history.listenBefore(() => {
+      if (this.shouldConfirmBeforeRoute()) {
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: This code is duplicated from above; Can you put this into a common method?

File: client/src/boot/BootRoutes.js [hunk 3]
```diff
@@ -102,9 +120,17 @@ class BootRoutes {
     const store = this.store;
 
     pageRouter('*', (ctx, next) => {
-      // eslint-disable-next-line no-param-reassign
-      ctx.store = store;
-      next();
+      if (this.shouldConfirmBeforeRoute()) {
+        if (window.confirm(i18n._t('Admin.CONFIRMUNSAVEDSHORT'))) {
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: If you re-shuffle your logic, you can combine these blocks.

```
if (!this.shouldConfirmBeforeRoute() || window.confirm(i18n._t('Admin.CONFIRMUNSAVEDSHORT')) {
	ctx.store = store;
	next();
} else {
	// whatever?
}
```

File: client/src/boot/BootRoutes.js [hunk 4]
```diff
@@ -140,8 +166,53 @@ class BootRoutes {
       });
     });
 
+    window.onbeforeunload = () => {
+      if (this.shouldConfirmBeforeUnload()) {
+        return i18n._t('Admin.CONFIRMUNSAVEDSHORT');
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: Third time this code is duplicated in this file. :)

**Comment by tractorcow** [comment]
Ok, this is weird... I can't even click links in the CMS anymore. I think you need another next() in your page.js callback?

**Comment by tractorcow** [comment]
image

**Comment by tractorcow** [comment]
Core change detection has broken in non-react sections.

- Go to pages section
- Edit any page
- Modify the title
- click "assets" section
- No change detection alert, you navigate immediately

I was expecting to see this alert:

image

**Comment by tractorcow** [comment]
This is the code which handles this (in LeftAndMain.EditForm.js)

```js
// Can't bind this through jQuery
window.onbeforeunload = function(e) {
  var form = $('.cms-edit-form');
  form.trigger('beforesubmitform');
  if(form.is('.changed') && ! form.is('.discardchanges')) {
    return i18n._t('Admin.CONFIRMUNSAVEDSHORT');
  }
};
```

File: client/src/boot/BootRoutes.js [hunk 5]
```diff
@@ -140,8 +156,64 @@ class BootRoutes {
       });
     });
 
+    window.onbeforeunload = this.handleBeforeUnload;
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: I think this is overwriting the `window.onbeforeunload` assigned via LeftAndMain.EditForm.js. See my comment.

**Comment by tractorcow** [comment]
Some more feedback:

When navigating in a react section I get this dialog:

image

When I navigate (hook based on window.onbeforeunload) I get this error.

image


Can we consolidate the messages to be the same?

When we use `window.confirm` we need to add an extra line to disclose the effects of "cancel" and "ok". note that this doesn't need to be exactly the same as the chrome default message (since that's added by the browser, and we have no control over it) but it needs to meet the same need.

**Comment by tractorcow** [comment]
Oh, it's already there.

use `Admin.CONFIRMUNSAVED` instead of `Admin.CONFIRMUNSAVEDSHORT`. Done :)

I'll push it up and see how it looks.

File: client/src/boot/BootRoutes.js [hunk 6]
```diff
@@ -140,8 +158,79 @@ class BootRoutes {
...
+   * with unsaved changes. Return `false` otherwise.
+   *
+   * @return {Boolean}
+   */
+  shouldConfirmBeforeRoute() {
+    const state = this.store.getState();
+    const unsaveds = state.unsavedForms || {};
+    const locationBeforeTransitions = state.routing.locationBeforeTransitions;
+    const pathname = locationBeforeTransitions && locationBeforeTransitions.pathname;
+    let ret = false;
```
Comment (flamerohr) [inline client/src/boot/BootRoutes.js]: @phalkunz for future reference, a meaningful variable name would be helpful, e.g. `shouldConfirm`

File: client/src/boot/BootRoutes.js [hunk 7]
```diff
@@ -140,8 +158,79 @@ class BootRoutes {
       });
     });
 
+    // Legacy support – make sure we don't overwrite the `onbeforeunload`
+    // handler in `LeftAndMain.EditForm.js` but run it after the new handler.
+    const currBeforeUnload = window.onbeforeunload;
+    window.onbeforeunload = () => {
+      if (this.shouldConfirmBeforeUnload()) {
```
Comment (flamerohr) [inline client/src/boot/BootRoutes.js]: a bit of code duplication here: e.g. `let message = this.handleBeforeUnload(); if (message) { return message; }`

File: client/src/boot/BootRoutes.js (hunk 6 above)
Comment (phalkunz) [inline client/src/boot/BootRoutes.js]: 👍

File: client/src/boot/BootRoutes.js [hunk 8]
```diff
@@ -140,8 +158,79 @@ class BootRoutes {
...
+
+  /**
+   * Return `true` if there are forms with unsaved changes. Return `false`
+   * otherwise.
+   *
+   * @return {Boolean}
+   */
+  shouldConfirmBeforeUnload() {
+    const state = this.store.getState();
+    const unsaveds = state.unsavedForms || {};
```
Comment (tractorcow) [inline client/src/boot/BootRoutes.js]: An oversight here is that not every unsaved form should block navigation if un-submitted; E.g. search forms.

**Pull request merged**